Extending Conan functionalities with hooks (#63)
conan-io/conan-io.github.io: _posts/2019-02-25-Extending-conan-with-hooks.markdown

---
layout: post
comments: false
title: "Extending Conan functionalities with hooks"
---

> Back in [Conan 1.8 blog post]https://blog.conan.io/2018/10/11/New-conan-release-1-8.html() release we introduced a so called

Reviewer: ```suggestion
Back in [Conan 1.8 blog post](https://blog.conan.io/2018/10/11/New-conan-release-1-8.html) release we introduced a so called
```

"Plugin System". Reading some of the feedback from users we soon realized that although it was a very useful feature, it wasn't exactly a
plugin mechanism. Normally such a mechanism is something more general and powerful that replaces or complements the functionality of a tool in a wider way.

Instead, the feature was designed with the philosophy on having a way of doing pre and post actions between certain Conan events in mind. This was very
similar to [git hooks feature](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks), so we decided to rename them after and consider
developing a real plugin system in the future.

> ## Status of Hooks

Reviewer 2: someone asked on Slack if hooks change the package id. Maybe it would be nice include this information on this blog post and update hooks docs.

Reviewer: Yes, that information/warning/recommendation is very relevant.

Author: That is explained in the Considerations section down below

> As you may know, hooks were released to give users a way perform additional tasks besides the usual actions in Conan like exporting,

Reviewer: ```suggestion
Hooks were released to give users a way to perform additional tasks besides the usual actions in Conan like exporting,
```

> building, uploading... and even to customize part of the Conan behavior. This view has not changed and we strongly believe they make a lot of

Reviewer: I would avoid the _customize part of the Conan behavior_, without modifying the `conanfile` object itself I think it is not possible to affect the Conan behavior (and I prefer not to see that kind of hooks).

...so I will change the mentions to _"custom behavior"_ with something like _"additional/extra behavior"_

Author: I was thinking about the metadata hook and similar ones that may arise

sense for people using Conan inside a company or organization that want a custom behavior when building all of their packages.

However, companies aren't the only ones that can benefit from the power of custom actions in Conan but also the open source community. Hooks
are very convenient when it comes to linting syntax in recipes, checking for missing attributes like licenses, ensuring proper application
of settings and so on.

In [recent releases](https://docs.conan.io/en/latest/changelog.html) we introduced some minor improvements and fixes for hooks although
usability remained almost the same. Hooks can now be installed in different folders under the *hooks* folder in the configuration, allowing

> users to have multiple hooks living together and avoiding naming collision. This structure may come handy when reusing modules in modules or

Reviewer: ```suggestion
users to have multiple hooks living together and avoiding naming collision. This structure may come handy when reusing modules in hooks or
```

storing additional files such as licenses, readmes, requirement files...

> Therefore, activation of hooks can now be done with a path to the hook interface:

Reviewer: ... relative to the `.conan/hooks` folder or even a full path to a hook somewhere else...

*conan.conf*
```
...

[hooks]
conanio/hooks/attribute_checker
conanio/hooks/conan-center
conanio/hooks/binary-linter
```

You can find more information about how to activate from command line and share hooks in the
[documentation](https://docs.conan.io/en/latest/extending/hooks.html#storage-activation-and-sharing).

## Hooks under development

In the documentation we proposed some ideas and small examples (https://docs.conan.io/en/latest/extending/hooks.html) about hooks like
attribute checks in the recipe, package signing or source code download immutability. We have already implemented the attribute checker
linter as a hook.

Together with this we created a hooks repository to start developing ourselves useful hooks for the community and we already got some pull
requests and bunch of issues with ideas on how to improve the hooks integration with Conan and discuss things like the configuration,
testing, documentation...

> Here we want to share the hooks created in this repo and how to use them.

Reviewer: "Here", "this repo" ---> I haven't seen any link to the repo yet.

### Cloning the hooks repository

As stated in the documentation, hooks can be shared with ``conan config install``, making them part of the configuration and managing also

> its default activation. This mechanism is useful for those sharing the configuration all together but, what do you do when want to try

Reviewer: "what if you are developing a hook?", "what if you want to have them versioned in their own repository?"

hooks developed here and there? The mechanism proposed for this in the documentation is using ``git clone`` directly in the *~/.conan/hooks*
directory using a subdirectory for them:

```

> $ git clone https://github.com/conan-io/hooks.git conan-io

Reviewer: Let's be sure that the `clone` is performed inside the hooks folder, for those users reading in diagonal

```
$ cd ~/.conan/hooks
$ git clone https://github.com/conan-io/hooks.git conan-io
```

```

This gives users the power of having hooks under a VCS and updating to new versions very simple.

Now it is just a matter of activating the desired hooks:

```
$ conan config set hooks.conanio/hooks/attribute_checker

> $ conan config set hooks.conanio/hooks/binary-linter

Reviewer: If the repo is cloned inside `conan-io` folder, it should be

```
$ conan config set hooks.conan-io/hooks/attribute_checker
$ conan config set hooks.conan-io/hooks/binary-linter
```

[HOOK - conanio/hooks/binary-linter.py] post_package(): checking file "C:\Users\danimtb\.conan\data\docopt\0.6.2\danitmb\testing\package\970e773c5651dc2560f86200a4ea56c23f568ff9\bin\docopt.dll"
[HOOK - conanio/hooks/binary-linter.py] post_package(): "C:\Users\danimtb\.conan\data\docopt\0.6.2\danitmb\testing\package\970e773c5651dc2560f86200a4ea56c23f568ff9\bin\docopt.dll" doesn't import library "msvcr110.dll"
...
[HOOK - conanio/hooks/binary-linter.py] post_package(): "C:\Users\danimtb\.conan\data\docopt\0.6.2\danitmb\testing\package\970e773c5651dc2560f86200a4ea56c23f568ff9\bin\docopt.dll" imports library "vcruntime140.dll"
[HOOK - conanio/hooks/binary-linter.py] post_package(): "C:\Users\danimtb\.conan\data\docopt\0.6.2\danitmb\testing\package\970e773c5651dc2560f86200a4ea56c23f568ff9\bin\docopt.dll" doesn't import library "vcruntime140d.dll"
...
```

See hook documentation: https://github.com/conan-io/hooks#binary-linter

### Bintray Updater

As some of you may know, when uploading packages to Bintray the metadata of the recipe is not process at all. This results in the

> information of Bintray being empty. However, with this hook you would get all the information filled.

Reviewer 2: Maybe you could add something more specific, like:

The information is collected from the recipe attributes, such as *name*, *license*, *url*, *homepage* and *description*. The maturity level is based on the branch name, names like `master`, `release` and `stable` are considered *Stable* maturity level. The project logo is not supported by this hook, since the Bintray API does not allow file uploads.

> You will have to provide your Bintray user and API token as environment variables (``CONAN_LOGIN_USERNAME`` & ``CONAN_PASSWORD``).

Reviewer 2: They need to be updated: https://github.com/conan-io/hooks/issues/36

> You will have to provide your Bintray user and API token as environment variables (``CONAN_LOGIN_USERNAME`` & ``CONAN_PASSWORD``).

Reviewer: ```suggestion
You will have to provide your Bintray user and API token as environment variables (``CONAN_LOGIN_USERNAME`` and ``CONAN_PASSWORD``).
```

TestPkg/0.0.1@user/channel: Generated conaninfo.txt
TestPkg/0.0.1@user/channel: Generated conanbuildinfo.txt
...
TestPkg/0.0.1@user/channel: Calling package()
TestPkg/0.0.1@user/channel: WARN: This conanfile has no package step
TestPkg/0.0.1@user/channel package(): WARN: No files in this package!
TestPkg/0.0.1@user/channel: Package 'ca33edce272a279b24f87dc0d4cf5bbdcffbc187' created
[HOOK - conanio/hooks/conan-center.py] post_package(): ERROR: [PACKAGE LICENSE] No package licenses found in: ~/.conan/data/TestPkg/0.0.1/user/channel/package/ca33edce272a279b24f87dc0d4cf5bbdcffbc187. Please package the library license to a 'licenses' folder
[HOOK - conanio/hooks/conan-center.py] post_package(): [DEFAULT PACKAGE LAYOUT] OK
[HOOK - conanio/hooks/conan-center.py] post_package(): [MATCHING CONFIGURATION] OK
[HOOK - conanio/hooks/conan-center.py] post_package(): [SHARED ARTIFACTS] OK
```

As you can see, all the checks are non blocking and mostly informative. There are recipe syntax checks and also license and binary format
ones.

See hook documentation: https://github.com/conan-io/hooks#conan-center

### GitHub Updater

> This hook is similar to the Bintray updater but with GitHub. It updates the repository information such as description, topics and webpage.

Reviewer 2: screenshot maybe?

![conan-boost](https://screenshotscdn.firefoxusercontent.com/images/42ac4ae5-1e90-46f7-80f3-d88975f875e1.png)

This way all the information is in GitHub and it will help when searching for recipe sources.

You will need to set a GitHub API token as environment variable and perform an export of the recipe you want to get its information updated.
It will use the URL attribute in the recipe to perform the update.

See hook documentation: https://github.com/conan-io/hooks#github-update

> ## Considerations

Reviewer 2: Include some comment about "How to include a new hook?" Not only the feedback from users, but maybe someone have some useful hook and it could be included as official.

---
layout: post
comments: false
title: "Extending Conan functionalities with hooks"
---

> Back in [Conan 1.8 blog post]https://blog.conan.io/2018/10/11/New-conan-release-1-8.html() release we introduced a so called

Reviewer 2: Really nice explaining why plugins is now called hooks. :+1:

"Plugin System". Reading some of the feedback from users we soon realized that although it was a very useful feature, it wasn't exactly a

> plugin mechanism. Normally a is something more general and powerful that replaces or complements the functionality of a tool in a wider way.

Author: Missing word "plugin"

> Instead, the feature was design in the philosophy on having a way of doing pre and post actions between certain Conan events. This was very

Author: was designed with

similar to [git hooks feature](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks), so we decided to rename them after and consider
developing a real plugin system in the future.

## Status of Hooks

> As you may know, hooks were release to give users a way perform additional tasks besides the usual actions in Conan like exporting,

Author: were released

> building, uploading... and even to customize part of the Conan behavior. This view has not change and we strongly believe they make a lot of

Author: has not changed

sense for people using Conan inside a company or organization that want a custom behavior when building all of their packages.

However, companies aren't the only ones that can benefit from the power of custom actions in Conan but also the open source community. Hooks
are very convenient when it comes to linting syntax in recipes, checking for missing attributes like licenses, ensuring proper application
of settings and so on.

In [recent releases](https://docs.conan.io/en/latest/changelog.html) we have introduced some minor improvements and fixes for hooks although

> usability remains almost the same. Hooks can now be installed in different folders under the *hooks* folder in the configuration, allowing

Author: remained

users to have multiple hooks living together and avoiding naming collision. This structure may come handy when reusing modules in modules or
storing additional files such as licenses, readmes, requirement files...

> Therefore, activation of hooks can now be done with a path to the hook interface (LINK TO DOCS):

Author: missing link

*conan.conf*
```
...

[hooks]
conanio/hooks/attribute_checker
conanio/hooks/conan-center
conanio/hooks/binary-linter
```

You can find more information about how to activate from command line and share hooks in the
[documentation](https://docs.conan.io/en/latest/extending/hooks.html#storage-activation-and-sharing).

## Hooks under development

In the documentation we proposed some ideas and small examples (https://docs.conan.io/en/latest/extending/hooks.html) about hooks like
attribute checks in the recipe, package signing or source code download immutability. We have already implemented the attribute checker
linter as a hook.

Together with this we created a hooks repository to start developing ourselves useful hooks for the community and we already got some pull
requests and bunch of issues with ideas on how to improve the hooks integration with Conan and discuss things like the configuration,
testing, documentation...

Here we want to share the hooks created in this repo and how to use them.

### Cloning the hooks repository

As stated in the documentation, hooks can be shared with ``conan config install``, making them part of the configuration and managing also

> its default activation. This mechanism is useful form people sharing the configuration all together but what do you do when want to try

Author: missing comma

$ conan create . docopt/0.6.2@user/testing
[HOOK - conanio/hooks/binary-linter.py] post_package(): conan binary linter plug-in
[HOOK - conanio/hooks/binary-linter.py] post_package(): file "C:\Users\danimtb\.conan\data\docopt\0.6.2\danitmb\testing\package\970e773c5651dc2560f86200a4ea56c23f568ff9\conaninfo.txt" is not a executable, skipping...
[HOOK - conanio/hooks/binary-linter.py] post_package(): checking file "C:\Users\danimtb\.conan\data\docopt\0.6.2\danitmb\testing\package\970e773c5651dc2560f86200a4ea56c23f568ff9\bin\docopt.dll"
[HOOK - conanio/hooks/binary-linter.py] post_package(): "C:\Users\danimtb\.conan\data\docopt\0.6.2\danitmb\testing\package\970e773c5651dc2560f86200a4ea56c23f568ff9\bin\docopt.dll" doesn't import library "msvcr110.dll"
...
[HOOK - conanio/hooks/binary-linter.py] post_package(): "C:\Users\danimtb\.conan\data\docopt\0.6.2\danitmb\testing\package\970e773c5651dc2560f86200a4ea56c23f568ff9\bin\docopt.dll" imports library "vcruntime140.dll"
[HOOK - conanio/hooks/binary-linter.py] post_package(): "C:\Users\danimtb\.conan\data\docopt\0.6.2\danitmb\testing\package\970e773c5651dc2560f86200a4ea56c23f568ff9\bin\docopt.dll" doesn't import library "vcruntime140d.dll"

> [HOOK - conanio/hooks/binary-linter.py] post_package(): "C:\Users\danimtb\.conan\data\docopt\0.6.2\danitmb\testing\package\970e773c5651dc2560f86200a4ea56c23f568ff9\bin\docopt.dll" imports library "vcruntime140.dll"

Reviewer 3: did it run twice? I mean, if it's a bug in hook or something else?

Author: Will remove that line, but yes, it is outputing that check against ``vcruntime140d.dll`` twice. It doesn't happen with the rest of the checks

for people using Conan inside a company or organization that want a custom behavior when building all of their packages.

However, companies aren't the only ones that can benefit from the power of custom actions in Conan but also the open source community. Hooks
are very convenient when it comes to linting syntax in recipes, checking for missing attributes like licenses, ensuring proper application
of settings and so on.

In [recent releases](https://docs.conan.io/en/latest/changelog.html) we introduced some minor improvements and fixes for hooks although
usability remained almost the same. Hooks can now be installed in different folders under the *hooks* folder in the configuration, allowing
users to have multiple hooks living together and avoiding naming collision. This structure may come handy when reusing modules in hooks or
storing additional files such as licenses, readmes, requirement files...

You can find more information about how to activate from command line and share hooks in the
[documentation](https://docs.conan.io/en/latest/extending/hooks.html#storage-activation-and-sharing).

## Using Hooks

> Hooks are just python scripts that could be more or less complicated but with an interface in common. To use them we would need to have the source code of the hook to be executed by Conan and get them activated with some commands.

Reviewer 2: ```suggestion
The Conan hooks is a python script intended to extend the Conan functionalities and let users customize the client behavior at determined points. To use them we would need to have the source code of the hook to be executed by Conan and get them activated with some commands.
```

---
layout: post
comments: false
title: "Extending Conan functionalities with hooks"
---

> Back in [Conan 1.8 blog post](https://blog.conan.io/2018/10/11/New-conan-release-1-8.html release we introduced a so called

Reviewer 2: ```suggestion
Back in [Conan 1.8 blog post](https://blog.conan.io/2018/10/11/New-conan-release-1-8.html) 
```

> Back in [Conan 1.8 blog post](https://blog.conan.io/2018/10/11/New-conan-release-1-8.html release we introduced a so called

Reviewer 2: ```suggestion
Back in [Conan 1.8 blog post](https://blog.conan.io/2018/10/11/New-conan-release-1-8.html) release we introduced a so called
```

make a lot of sense for people using Conan inside a company or organization that want a custom behavior when building all of their packages.

However, companies aren't the only ones that can benefit from the power of custom actions in Conan but also the open source community. Hooks
are very convenient when it comes to linting syntax in recipes, checking for missing attributes like licenses, ensuring proper application
of settings and so on.

In [recent releases](https://docs.conan.io/en/latest/changelog.html) we introduced some minor improvements and fixes for hooks although
usability remained almost the same. Hooks can now be installed in different folders under the *hooks* folder in the configuration, allowing
users to have multiple hooks living together and avoiding naming collision. This structure may come handy when reusing modules in hooks or
storing additional files such as licenses, readmes, requirement files...

You can find more information about how to activate from command line and share hooks in the
[documentation](https://docs.conan.io/en/latest/extending/hooks.html#storage-activation-and-sharing).

## Using Hooks

> The Conan hooks are a Python scripts intended to extend the Conan functionalities and let users enhance the client behavior at determined

Reviewer 3: I think hooks aren't `python scripts`, as they aren't intended to be executed directly from the shell, right? they are intended to be imported, so they are more likely `python modules` or something like that

Author: ok

> points. To use them we would need to have the source code of the hook to be executed by Conan and get them activated with some commands.

Reviewer 3: would be nice to add a few examples of such points here (export, upload, etc)

Author: Well, there are links all over the post referecing the interface and the documentation, but ok

Let's take a look to how to do this.

### Cloning a Hooks repository

As stated in the documentation, hooks can be shared with ``conan config install``, making them part of the configuration and managing also

> its default activation. This mechanism is useful for those sharing the configuration all together but, what if you are developing a hook?,

Reviewer 3: their default activation?

Author: I meant the activation comming directly as part of the conan.conf

what if you want to have them versioned in their own repository? The mechanism proposed for this in the documentation is using ``git clone``
directly in the *~/.conan/hooks* directory using a subdirectory for them:

```
$ cd ~/.conan/hooks
$ git clone https://github.com/conan-io/hooks.git conan-io
```

> We strongly believe they need a separated repository to grow and get mature, so weWe have come with the solution on

Reviewer 2: ```suggestion
We strongly believe they need a separated repository to grow and get mature, so we have come with the solution on
```

having the development controlled under ``git`` and this will help us with the versioning in the future.

You might have noticed that we cloned the repository into a ``conan-io`` folder, this is done to avoid name collision of hooks so we can
distinguish them using its path.

### Activating some Hooks

Now that we have hooks cloned it is just a matter of activating the desired ones. There are different possibilities for doing this:

> - Editing the *conan.conf* file: You can go to the ``[hooks]`` section and write the path to the hook you want to use relative to the

Reviewer 3: I personally don't think it's good to propose direct editing of config files if command line method is available, this is more like implementation detail which may change

Author: Editing the *conan.conf* by hand is something quite common. I am proposing all the possibilities

  *hooks* folder.

  *conan.conf*
  ```
  ...

  [hooks]
  conan-io/hooks/attribute_checker
  conan-io/hooks/binary_linter
  ```

- From command line: Using `conan config set hooks.<hook path>` command (relative path as described above).

  ```
  $ conan config set hooks.conan-io/hooks/attribute_checker
  $ conan config set hooks.conan-io/hooks/binary_linter
  ...
  ```

- Using the environment variable ``CONAN_HOOKS``: This allows using more than one hook separated by commas and path can be either relative
  to *hooks* folder or absolute to any other location. This might be useful for CI environments.

  ```
  $ set CONAN_HOOKS=/home/user/.conan/hooks/conan-io/hooks/attribute_checker,/home/user/.conan/hooks/conan-io/hooks/binary_linter
  ```

## Hooks under development

> In the documentation we proposed some ideas and small examples (https://docs.conan.io/en/latest/extending/hooks.html) about hooks like

Reviewer 2: ```suggestion
In the documentation we proposed some ideas and small [examples](https://docs.conan.io/en/latest/extending/hooks.html) about hooks like
```

attribute checks in the recipe, package signing or source code download immutability. We have already implemented the attribute checker
linter as a hook.

Together with this we created a hooks repository to start developing ourselves useful hooks for the community and we already got some pull
requests and bunch of issues with ideas on how to improve the hooks integration with Conan and discuss things like the configuration,
testing, documentation...

In this post we want to share the hooks created in this repo and how to use them.

### Attribute Checker

> This was the first hook created as an example and is now extracted in this repository. It is a small example to learn how to write a first

Reviewer 2: As the command line example is too long to read, I would say to explain what the purpose of this hook is. We can roll easily on computer, but reading in a smartphone it's not comfortable I think. 

![](https://screenshotscdn.firefoxusercontent.com/images/2ac97f14-bc3d-49b1-9349-c8f592103a2f.png)

Author: I think having the real output is important

Reviewer 2: I'm not saying to remove the output, just to add some comment like "it checks if  important attributes are present in the recipe.". All other hooks on this post have some brief description.

hook.

```
$ conan export . user/channel
[HOOK - conan-io/hooks/attribute_checker.py] pre_export(): WARN: Conanfile doesn't have 'url'. It is recommended to add it as attribute
[HOOK - conan-io/hooks/attribute_checker.py] pre_export(): WARN: Conanfile doesn't have 'license'. It is recommended to add it as attribute
```

> See hook documentation: https://github.com/conan-io/hooks#attribute-checker

Reviewer 2: ```suggestion
See hook documentation: [attribute-checker](https://github.com/conan-io/hooks#attribute-checker)
```

The link is not automatically solved:
![](https://screenshotscdn.firefoxusercontent.com/images/9f05ceec-3286-4471-b8fd-66adcf6e76fc.png)

Author: thanks

profile used.

Here you can see a brief output of the hook:

```
$ conan create . docopt/0.6.2@user/testing
[HOOK - cona-nio/hooks/binary_linter.py] post_package(): conan binary linter plug-in
[HOOK - conan-io/hooks/binary_linter.py] post_package(): file "~/.conan/data/docopt/0.6.2/user/testing/package/970e773c5651dc2560f86200a4ea56c23f568ff9/conaninfo.txt" is not a executable, skipping...
[HOOK - conan-io/hooks/binary_linter.py] post_package(): checking file "~/.conan/data/docopt/0.6.2/user/testing/package/970e773c5651dc2560f86200a4ea56c23f568ff9/bin/docopt.dll"
[HOOK - conan-io/hooks/binary_linter.py] post_package(): "~/.conan/data/docopt/0.6.2/user/testing/package/970e773c5651dc2560f86200a4ea56c23f568ff9/bin/docopt.dll" doesn't import library "msvcr110.dll"
...
[HOOK - conan-io/hooks/binary_linter.py] post_package(): "~/.conan/data/docopt/0.6.2/user/testing/package/970e773c5651dc2560f86200a4ea56c23f568ff9/bin/docopt.dll" imports library "vcruntime140.dll"
[HOOK - conan-io/hooks/binary_linter.py] post_package(): "~/.conan/data/docopt/0.6.2/user/testing/package/970e773c5651dc2560f86200a4ea56c23f568ff9/bin/docopt.dll" doesn't import library "vcruntime140d.dll"
...
```

> See hook documentation: https://github.com/conan-io/hooks#binary-linter

Reviewer 2: ```suggestion
See hook documentation: [binary-linter](https://github.com/conan-io/hooks#binary-linter)
```

You will have to provide your Bintray user and API token as environment variables (``BINTRAY_LOGIN_USERNAME`` and ``BINTRAY_PASSWORD``).

With this hook active, the information is collected from the recipe attributes, such as ``name``, ``license``, ``url``, ``homepage`` and
``description``. The maturity level is based on the branch name like `master`, `release` and `stable` are considered ``Stable`` maturity
level. The project logo is not supported by this hook, since the Bintray API does not allow file uploads, and it is updated during the
recipe upload using the Bintray REST API (LINK).

```
$ conan upload docopt/0.6.2@user/testing -r bintray

Uploading docopt/0.6.2@user/testing to remote 'bintray'
...
Uploaded conan recipe 'docopt/0.6.2@user/testing' to 'bintray': https://bintray.com/user/public-conan
[HOOK - conan-io/hooks/bintray_update.py] post_upload_recipe(): Reading package info form Bintray...
[HOOK - conan-io/hooks/bintray_update.py] post_upload_recipe(): Inspecting recipe info ...
[HOOK - conan-io/hooks/bintray_update.py] post_upload_recipe(): Bintray is outdated. Updating Bintray package info: licenses, issue_tracker_url, vcs_url, website_url, desc
```

> See hook documentation: https://github.com/conan-io/hooks#bintray-update

Reviewer 2: ```suggestion
See hook documentation: [bintray-update](https://github.com/conan-io/hooks#bintray-update)
```

[HOOK - conan-io/hooks/conan-center_reviewer.py] post_build(): [SHARED ARTIFACTS] OK
TestPkg/0.0.1@user/channel: Generated conaninfo.txt
TestPkg/0.0.1@user/channel: Generated conanbuildinfo.txt
...
TestPkg/0.0.1@user/channel: Calling package()
TestPkg/0.0.1@user/channel: WARN: This conanfile has no package step
TestPkg/0.0.1@user/channel package(): WARN: No files in this package!
TestPkg/0.0.1@user/channel: Package 'ca33edce272a279b24f87dc0d4cf5bbdcffbc187' created
[HOOK - conan-io/hooks/conan-center_reviewer.py] post_package(): ERROR: [PACKAGE LICENSE] No package licenses found in: ~/.conan/data/TestPkg/0.0.1/user/channel/package/ca33edce272a279b24f87dc0d4cf5bbdcffbc187. Please package the library license to a 'licenses' folder
[HOOK - conan-io/hooks/conan-center_reviewer.py] post_package(): [DEFAULT PACKAGE LAYOUT] OK
[HOOK - conan-io/hooks/conan-center_reviewer.py] post_package(): [MATCHING CONFIGURATION] OK
[HOOK - conan-io/hooks/conan-center_reviewer.py] post_package(): [SHARED ARTIFACTS] OK
```

As you can see, all the checks are non blocking and mostly informative. There are recipe syntax checks and also license and binary format
ones.

> See hook documentation: https://github.com/conan-io/hooks#conan-center

Reviewer 2: ```suggestion
See hook documentation: [conan-center_reviewer](https://github.com/conan-io/hooks#conan-center_reviewer)
```

### GitHub Updater

This hook is similar to the Bintray updater but with GitHub. It updates the repository information such as description, topics and webpage.
This way all the information is in GitHub and it will help when searching for recipe sources.

> ![](../assests/post_images/2019-02-27/github_updater.png)

Reviewer 2: ```suggestion
<p class="centered">
    <img src="{{ site.url }}/assets/post_images/2019-02-27/github_updater.png" align="center"/>
</p>
```
Relative path does not work:

![](https://screenshotscdn.firefoxusercontent.com/images/bb1579d6-da01-45e2-8429-3a2e013683cc.png)

> You will need to set a GitHub API token as environment variable and perform an export of the recipe you want to get its information updated.

Reviewer 2: ```suggestion
You will need to set a GitHub API token as environment variable ``GITHUB_TOKEN`` and perform an export of the recipe you want to get its information updated.
```

It will use the URL attribute in the recipe to perform the update.

> See hook documentation: https://github.com/conan-io/hooks#github-updater

Reviewer 2: ```suggestion
See hook documentation: [github-updater](https://github.com/conan-io/hooks#github-updater)
```

## Considerations for Hooks development

> Keep in mind that hooks are quite versatile and can be use as far as python extends, however, it is not recommended to use hooks for task

Reviewer 3: can be used?

```

See hook documentation: [attribute_checker](https://github.com/conan-io/hooks#attribute-checker)

### Binary Linter

The binary linter hooks provides some hints about the artifacts that has been built and eventually packaged after the ``package()`` call. It
is very helpful to analyze possible missing dependencies in shared libraries as well as to check that the binaries generated match the
profile used.

Here you can see a brief output of the hook:

```
$ conan create . docopt/0.6.2@user/testing
[HOOK - cona-nio/hooks/binary_linter.py] post_package(): conan binary linter plug-in
[HOOK - conan-io/hooks/binary_linter.py] post_package(): file "~/.conan/data/docopt/0.6.2/user/testing/package/970e773c5651dc2560f86200a4ea56c23f568ff9/conaninfo.txt" is not a executable, skipping...
[HOOK - conan-io/hooks/binary_linter.py] post_package(): checking file "~/.conan/data/docopt/0.6.2/user/testing/package/970e773c5651dc2560f86200a4ea56c23f568ff9/bin/docopt.dll"
[HOOK - conan-io/hooks/binary_linter.py] post_package(): "~/.conan/data/docopt/0.6.2/user/testing/package/970e773c5651dc2560f86200a4ea56c23f568ff9/bin/docopt.dll" doesn't import library "msvcr110.dll"
...

> [HOOK - conan-io/hooks/binary_linter.py] post_package(): "~/.conan/data/docopt/0.6.2/user/testing/package/970e773c5651dc2560f86200a4ea56c23f568ff9/bin/docopt.dll" imports library "vcruntime140.dll"

Reviewer 4: This is nice, but I'd expect from this linter to output warning messages if these runtimes doesn't match the ``settings.compiler.runtime`` defined in conan settings.

Nothing that should be changed in this blog post, just an idea for the hook

Author: cc/ @SSE4

[HOOK - conan-io/hooks/binary_linter.py] post_package(): "~/.conan/data/docopt/0.6.2/user/testing/package/970e773c5651dc2560f86200a4ea56c23f568ff9/bin/docopt.dll" doesn't import library "vcruntime140d.dll"
...
```

See hook documentation: [binary_linter](https://github.com/conan-io/hooks#binary-linter)

### Bintray Updater

> As some of you may know, when uploading packages to Bintray the metadata of the recipe is not process at all. This results in the

Reviewer 4: is not processed at all

information of Bintray being empty. However, with this hook you would get all the information filled.

You will have to provide your Bintray user and API token as environment variables (``BINTRAY_LOGIN_USERNAME`` and ``BINTRAY_PASSWORD``).

With this hook active, the information is collected from the recipe attributes, such as ``name``, ``license``, ``url``, ``homepage`` and
``description``. The maturity level is based on the branch name like `master`, `release` and `stable` are considered ``Stable`` maturity
level. The project logo is not supported by this hook, since the Bintray API does not allow file uploads, and it is updated during the

> recipe upload using the Bintray REST API (LINK).

Reviewer 4: (LINK)

[HOOK - conan-io/hooks/conan-center_reviewer.py] post_package(): [MATCHING CONFIGURATION] OK
[HOOK - conan-io/hooks/conan-center_reviewer.py] post_package(): [SHARED ARTIFACTS] OK
```

As you can see, all the checks are non blocking and mostly informative. There are recipe syntax checks and also license and binary format
ones.

See hook documentation: [conan-center_reviewer](https://github.com/conan-io/hooks#conan-center-reviewer)

### GitHub Updater

This hook is similar to the Bintray updater but with GitHub. It updates the repository information such as description, topics and webpage.
This way all the information is in GitHub and it will help when searching for recipe sources.

<p class="centered">
    <img src="{{ site.url }}/assets/post_images/2019-02-27/github_updater.png" align="center"/>
</p>

You will need to set a GitHub API token as environment variable ``GITHUB_TOKEN`` and perform an export of the recipe you want to get its information updated.
It will use the URL attribute in the recipe to perform the update.

See hook documentation: [github-updater](https://github.com/conan-io/hooks#github-updater)

## Considerations for Hooks development

> Keep in mind that hooks are quite versatile and can be used as far as python extends, however, it is not recommended to use hooks for task

Reviewer 4: for tasks
